File: Python/singleNumber3.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: singleNumber3.py
@Function: 只出现一次的数字 III
@Link: https://leetcode-cn.com/problems/single-number-iii/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-22
"""


from typing import List
import functools


class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        ret = functools.reduce(lambda x, y: x ^ y, nums)
        div = 1
        while div & ret == 0:
            div <<= 1
        a, b = 0, 0
        for n in nums:
            if n & div:
                a ^= n
            else:
                b ^= n
        return [a, b]


# ------------------------
if __name__ == "__main__":
    nums = [1,2,1,3,2,5]

    solution = Solution()
    single_nums = solution.singleNumber(nums)
    print(f"The solution of this problem is {single_nums}")

File: C++/singleNumber3.cpp
/**
 * @File    : singleNumber3.cpp
 * @Brief   : 只出现一次的数字 III
 * @Link    : https://leetcode-cn.com/problems/single-number-iii/
 * @Author  : Wei Li
 * @Date    : 2021-07-22
*/

#include <iostream>
#include <vector>

class Solution
{
public:
    std::vector<int> singleNumber(std::vector<int> &nums)
    {
        int ret = 0;
        for (int n : nums)
            ret ^= n;
        int div = 1;
        while ((div & ret) == 0)
            div <<= 1;
        int a = 0, b = 0;
        for (int n : nums)
            if (div & n)
                a ^= n;
            else
                b ^= n;
        return std::vector<int>{a, b};
    }
};

void DisplayElement(const std::vector<int> & container)
{
    for (const auto element : container)
    {
        std::cout << element << " ";
    }
    std::cout << std::endl;
    
}

// -----------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums{1, 2, 1, 3, 2, 5};

    Solution solution;
    auto single_nums = solution.singleNumber(nums);
    DisplayElement(single_nums);

    return 0;
}
